File: integration/feature/test-quick/resources/app/src/MyNumber.scala
package app

import lib.*

final case class MyNumber(val value: Int)

object MyNumber {
  
  given gCombinator: Combinator[MyNumber] = new Combinator[MyNumber] {
    def combine(a: MyNumber, b: MyNumber): MyNumber = MyNumber(a.value + b.value)
  }

  given gDefaultValue: DefaultValue[MyNumber] = new DefaultValue[MyNumber] {
    def defaultValue: MyNumber = MyNumber(0)
  }

  def combine(a: MyNumber, b: MyNumber, c: MyNumber): MyNumber = gCombinator.combine2(a, b, c)

  def defaultValue: MyNumber = gDefaultValue.defaultValue

}

File: integration/feature/test-quick-java/resources/app/src/MyNumber.java
package app;

import lib.Combinator;
import lib.DefaultValue;

public class MyNumber {
    private final int value;

    public MyNumber(int value) {
        this.value = value;
    }

    public int getValue() {
        return value;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof MyNumber) {
            return value == ((MyNumber) obj).value;
        }
        return false;
    }

    @Override
    public int hashCode() {
        return Integer.hashCode(value);
    }

    @Override
    public String toString() {
        return "MyNumber(" + value + ")";
    }

    public static final Combinator<MyNumber> COMBINATOR = new Combinator<MyNumber>() {
        @Override
        public MyNumber combine(MyNumber a, MyNumber b) {
            return new MyNumber(a.getValue() + b.getValue());
        }
    };

    public static final DefaultValue<MyNumber> DEFAULT_VALUE = new DefaultValue<MyNumber>() {
        @Override
        public MyNumber defaultValue() {
            return new MyNumber(0);
        }
    };

    public static MyNumber combine(MyNumber a, MyNumber b, MyNumber c) {
        return COMBINATOR.combine2(a, b, c);
    }

    public static MyNumber getDefaultValue() {
        return DEFAULT_VALUE.defaultValue();
    }
}
